### DISCUSSION TITLE
How to setup a test resource correctly to read it in a Scala CLI project

### DISCUSSION THREAD

**Discussion by halloleo** [VirtusLab/scala-cli#3138] [Using directives and cmd configuration options]
I use Scala CLI (and MUnit) in a small Scala 3 project

In a test I want to read a resource. Under the project root the resource is at the path:

    src/test/resources/testResource.txt

The Scala file of the test to get the resource is at 

    src/test/scala/org/myorg/myproject/mypack/App.test.scala

And the file looks like this:

```
//> using test.dep org.scalameta::munit::1.0.0

package org.myorg.myproject.mypack.tests

import java.io.InputStream

class ResourceTests extends munit.FunSuite: 
  test("getResource"):
    val inp: InputStream = getClass.getResourceAsStream("/testResource.txt")
    assertNotEquals(inp, null)
```

When I run the Scala CLI command

    scala-cli test .

it picks up the test and runs it, but the test always fails, beacuse `inp` is null. So the test does not find the resource. 

I have tried as resource location `"testResource.txt"` (so without the leading slash) - to no avail.

I also  have tried to tell Scala CLI that `src/test/resources` is a resource directory by adding

```
//> using test.resourceDir src/test/resources 
```

to `App.test.scala`, but the test results stay the same too.

So, how can I correct this? How can my test find the resource when using Scala CLI's `test` command? Do I have to put the resource somewhere else?

Thank you very much for any pointers!

**Comment by Gedochao** [comment]
Not sure what could the problem be... here's what I just tested and works fine:

```scala
//> using test.dep org.scalameta::munit::1.0.1
//> using test.resourceDir ./resources
package org.myorg.myproject.mypack.tests

import java.io.InputStream

class ResourceTests extends munit.FunSuite:
  test("getResource"):
    val inp: InputStream = getClass.getResourceAsStream("/testResource.txt")
    assertNotEquals(inp, null)
```

```bash
tree
# .
# ├── Example.test.scala
# └── resources
#     └── testResource.txt
# 
# 2 directories, 2 files
scala-cli test .
# Compiling project (test, Scala 3.5.0, JVM (17))
# Compiled project (test, Scala 3.5.0, JVM (17))
# org.myorg.myproject.mypack.tests.ResourceTests:
#   + getResource 0.006s
```

**Reply by halloleo** [reply]
@Gedochao Thanks for chiming in and checking this out.

You are right. It works -- but only if the resource directory _is in a subdirectory_ from the Scala test file.

However I use the standard Gradle/Maven layout like this (using your ):

```
.
└── src
    ├── main
    └── test
        ├── resources
        │   └── testResource.txt
        └── scala
            └── org
                └── myorg
                    └── myproject
                        └── mypack
                            └── Example.test.scala
```

Then, as I understand now, only the directive

    //> using test.resourceDir ../../../../../resources

works. -- And this sort of defeats the propose of using `getResource` (over reading from a full file path).

In Java it did work to put the resource in the standard location. Is there a way to achieve this with Scala CLI?

**Reply by Gedochao** [reply]
> In Java it did work to put the resource in the standard location.

What do you mean by a standard location in Java? You can use Java with Scala CLI, it will work exactly the same.
There is no default location for the resources, if that's what you mean, you do have to specify the resource directory.

> It works -- but only if the resource directory is in a _subdirectory_ from the Scala test file.

Yeah, if you'd like to pass an individual resource file rather than a directory, there is an issue for it (and a tied PR in progress):
- https://github.com/VirtusLab/scala-cli/issues/792

**Reply by halloleo** [reply]
Sorry, I wasn't clear enough: The Java projects were we have resources in the locations like 

```
src/main/scala/...
src/test/scala/...
src/test/resources
```

use gradle/maven, so I guess the resource path magic is done by these build systems.

Scala CLI needs the exact location specified, right?

**Reply by Gedochao** [reply]
> Scala CLI needs the exact location specified, right?

Yes. We don't have default locations for sources, resources, or any other inputs, really.
